drivers/clock_control/clock_control_ll_nrfxlib.c

```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <soc.h>
+#include <errno.h>
+#include <device.h>
+#include <kernel_includes.h>
+#include <clock_control.h>
+#include <ble_controller_soc.h>
+#include <multithreading_lock.h>
+
+static int hf_clock_start(struct device *dev, clock_control_subsys_t sub_system)
+{
+	ARG_UNUSED(dev);
+
+	s32_t errcode = MULTITHREADING_LOCK_ACQUIRE();
+
+	if (errcode == 0) {
+		errcode = ble_controller_hf_clock_request(NULL);
+		MULTITHREADING_LOCK_RELEASE();
+	}
+	if (errcode != 0) {
+		return -EFAULT;
+	}
+
+	bool blocking = POINTER_TO_UINT(sub_system);
+
+	if (blocking) {
+		bool is_running = false;
+
+		while (!is_running) {
+			errcode = MULTITHREADING_LOCK_ACQUIRE();
+			if (errcode == 0) {
+				errcode = ble_controller_hf_clock_is_running(
+					&is_running);
+				MULTITHREADING_LOCK_RELEASE();
+			}
+			if (errcode != 0) {
+				return -EFAULT;
+			}
+		}
+	}
+
+	return 0;
+}
+
+static int hf_clock_stop(struct device *dev, clock_control_subsys_t sub_system)
+{
+	ARG_UNUSED(dev);
+	ARG_UNUSED(sub_system);
+
+	s32_t errcode = MULTITHREADING_LOCK_ACQUIRE();
+
+	if (errcode == 0) {
+		errcode = ble_controller_hf_clock_release();
+		MULTITHREADING_LOCK_RELEASE();
+	}
+	if (errcode != 0) {
+		return -EFAULT;
+	}
+
+	return 0;
+}
+
+static int hf_clock_get_rate(struct device *dev,
+			     clock_control_subsys_t sub_system, u32_t *rate)
+{
+	ARG_UNUSED(dev);
+	ARG_UNUSED(sub_system);
+
+	if (rate == NULL) {
+		return -EINVAL;
+	}
+
+	*rate = MHZ(16);
+	return 0;
+}
+
+static int lf_clock_start(struct device *dev, clock_control_subsys_t sub_system)
+{
+	ARG_UNUSED(dev);
+	ARG_UNUSED(sub_system);
+
+	/* No-op. LFCLFK is started by default. */
```

Comment:
But not before BLE stack initialization as far as I understand.
What if somebody would like to manage clock before starting Bluetooth?
Please note that system clock driver (which is initialized before kernel and high-level services) requests LFCLK.